File: src/notification_tray/components/notification_service.py
import logging
from datetime import UTC, datetime
from enum import IntEnum
from pathlib import Path

import dbus
import dbus.decorators
import dbus.service
from PyQt5.QtCore import QObject, pyqtSignal

from notification_tray.utils.logging import log_input_and_output
from notification_tray.utils.paths import get_output_path

from ..types.notification import (CachedNotification, Notification,
                                  NotificationHints)

logger = logging.getLogger(__name__)
VERSION = "0.0.1"


class NotificationCloseReason(IntEnum):
    EXPIRED = 1
    DISMISSED_BY_USER = 2
    CLOSED_BY_CALL_TO_CLOSENOTIFICATION = 3
    UNDEFINED = 4


class NotificationServiceSignaler(QObject):
    notification_ready = pyqtSignal(int)
    notification_closed = pyqtSignal(int, int)


class NotificationService(dbus.service.Object):
    notifications = dict[int, CachedNotification]()

    def __init__(self, root_path: Path):
        bus_name = dbus.service.BusName(
            "org.freedesktop.Notifications", bus=dbus.SessionBus()
        )
        super().__init__(bus_name, "/org/freedesktop/Notifications")
        self.root_path = root_path
        self.signaler = NotificationServiceSignaler()
        logger.info(f"Started notification service with root_path {root_path}")

    @log_input_and_output(logging.INFO)
    @dbus.decorators.signal(  # type: ignore
        dbus_interface="org.freedesktop.Notifications", signature="us"
    )
    def ActionInvoked(self, id: int, action_key: str): ...

    @log_input_and_output(logging.INFO)
    @dbus.decorators.signal(  # type: ignore
        dbus_interface="org.freedesktop.Notifications", signature="usss"
    )
    def NotificationDisplayed(
        self,
        id: int,
        app_name: str,
        summary: str,
        body: str,
    ): ...

    @log_input_and_output(logging.INFO)
    @dbus.decorators.signal(  # type: ignore
        dbus_interface="org.freedesktop.Notifications", signature="uu"
    )
    def NotificationClosed(self, id: int, reason: NotificationCloseReason):
        pass

    @dbus.decorators.method(  # type: ignore
        "org.freedesktop.Notifications",
        in_signature="susssasa{sv}i",
        out_signature="u",
    )
    def Notify(
        self,
        app_name: str,
        replaces_id: int,
        app_icon: str,
        summary: str,
        body: str,
        actions: list[str],
        hints: NotificationHints,
        expire_timeout: int,
    ):
        # from https://specifications.freedesktop.org/notification-spec/latest/protocol.html: Servers must make sure not to return zero as an ID.
        logger.info(f"Got notification from {app_name} with summary {summary}")
        id = replaces_id or len(self.notifications) + 1
        logger.info(f"Notification ID: {id}")
        notification = Notification(
            app_name=app_name,
            replaces_id=replaces_id,
            app_icon=app_icon,
            summary=summary,
            body=body,
            expire_timeout=expire_timeout,
            id=id,
            actions=dict(zip(actions[::2], actions[1::2])),
            hints=hints,
            at=datetime.now(UTC),
        )
        self.notifications[id] = CachedNotification(
            **notification, path=get_output_path(self.root_path, notification)
        )
        self.signaler.notification_ready.emit(id)
        logger.info(f"Notification Ready. ID: {id}")
        return id

    @log_input_and_output(logging.INFO)
    @dbus.decorators.method(  # type: ignore
        "org.freedesktop.Notifications",
        in_signature="u",
        out_signature="",
    )
    def CloseNotification(self, id: int):
        if id in self.notifications and not self.notifications[id].get("trashed"):
            self.notifications[id]["closed_at"] = datetime.now(UTC)
            reason = NotificationCloseReason.CLOSED_BY_CALL_TO_CLOSENOTIFICATION
            self.NotificationClosed(id, reason)
            self.signaler.notification_closed.emit(id, reason)

        else:
            raise dbus.exceptions.DBusException()

    @log_input_and_output(logging.INFO)
    @dbus.decorators.method(  # type: ignore
        "com.github.NotificationTray",
        in_signature="",
        out_signature="",
    )
    def CloseActiveNotifications(self):
        for id, notification in self.notifications.items():
            if not notification.get("closed_at"):
                self.CloseNotification(id)

    @log_input_and_output(logging.INFO)
    @dbus.decorators.method(  # type: ignore
        "com.github.NotificationTray",
        in_signature="",
        out_signature="",
    )
    def OpenActiveNotifications(self):
        ids = list[tuple[int, str]]()
        for id, notification in self.notifications.items():
            if not notification.get("closed_at"):
                actions = notification["actions"]
                match len(actions):
                    case 0:
                        pass
                    case 1:
                        ids.append((id, next(iter(actions))))
                    case _:
                        raise dbus.exceptions.DBusException(f"Notification id {id} has more than one action")

        for id, action_key in ids:
            self.ActionInvoked(id, action_key)

    @dbus.decorators.method("org.freedesktop.Notifications", out_signature="ssss")  # type: ignore
    def GetServerInformation(self) -> tuple[str, str, str, str]:
        return ("notification-tray", "github.com", VERSION, "1.3")

    @dbus.decorators.method("org.freedesktop.Notifications", out_signature="as")  # type: ignore
    def GetCapabilities(self):
        return [
            "action-icons",
            "actions",
            "body",
            "body-hyperlinks",
            "body-images",
            "body-markup",
            "persistence",
            "sound",
        ]

File: src/notification_service.cpp
#include "notification_service.h"

#include "utils/logging.h"
#include "utils/paths.h"

#include <QDBusConnection>
#include <QDBusMessage>

static Logger logger = Logger::getLogger("NotificationService");

NotificationService::NotificationService(const fs::path& root_path, const QString& run_id,
                                         QObject* parent)
    : QObject(parent)
    , m_root_path(root_path)
    , m_run_id(run_id) {
    signaler = new NotificationServiceSignaler(this);

    QDBusConnection dbus = QDBusConnection::sessionBus();
    if (!dbus.registerService("org.freedesktop.Notifications")) {
        logger.error("Failed to register D-Bus service");
    }
    if (!dbus.registerObject("/org/freedesktop/Notifications", this,
                             QDBusConnection::ExportScriptableContents)) {
        logger.error("Failed to register D-Bus object");
    }

    logger.info(QString("Started notification service with root_path %1")
                    .arg(QString::fromStdString(root_path.string())));
}

uint NotificationService::Notify(const QString& app_name, uint replaces_id, const QString& app_icon,
                                 const QString& summary, const QString& body,
                                 const QStringList& actions, const QVariantMap& hints,
                                 int expire_timeout) {
    logger.info(QString("Got notification from %1 with summary %2").arg(app_name, summary));

    uint id = replaces_id ? replaces_id : static_cast<uint>(notifications.size() + 1);
    logger.info(QString("Notification ID: %1").arg(id));

    Notification notification;
    notification.app_name = app_name;
    notification.replaces_id = replaces_id;
    notification.app_icon = app_icon;
    notification.summary = summary;
    notification.body = body;
    notification.expire_timeout = expire_timeout;
    notification.id = id;
    notification.hints = hints;
    notification.at = QDateTime::currentDateTimeUtc();
    notification.notification_tray_run_id = m_run_id;

    for (int i = 0; i < actions.size(); i += 2) {
        if (i + 1 < actions.size()) {
            notification.actions[actions[i]] = actions[i + 1];
        }
    }

    CachedNotification cached;
    cached.app_name = notification.app_name;
    cached.replaces_id = notification.replaces_id;
    cached.app_icon = notification.app_icon;
    cached.summary = notification.summary;
    cached.body = notification.body;
    cached.expire_timeout = notification.expire_timeout;
    cached.id = notification.id;
    cached.actions = notification.actions;
    cached.hints = notification.hints;
    cached.at = notification.at;
    cached.notification_tray_run_id = notification.notification_tray_run_id;
    cached.path = Paths::getOutputPath(m_root_path, notification);

    notifications[id] = cached;
    emit signaler->notificationReady(id);

    logger.info(QString("Notification Ready. ID: %1").arg(id));
    return id;
}

void NotificationService::CloseNotification(uint id) {
    logger.info(QString("CloseNotification called for ID: %1").arg(id));

    auto it = notifications.find(id);
    if (it != notifications.end()) {
        it->second.closed_at = QDateTime::currentDateTimeUtc();
        if (!it->second.trashed) {
            auto reason =
                static_cast<uint>(NotificationCloseReason::CLOSED_BY_CALL_TO_CLOSENOTIFICATION);
            emit NotificationClosed(id, reason);
            emit signaler->notificationClosed(id, reason);
        }
    } else {
        if (calledFromDBus()) {
            sendErrorReply(QDBusError::InvalidArgs,
                           QString("Notification with id %1 not found").arg(id));
        }
        logger.error(QString("CloseNotification: ID %1 not found").arg(id));
    }
}

void NotificationService::CloseActiveNotifications() {
    logger.info("CloseActiveNotifications called");
    for (auto& [id, notification] : notifications) {
        // Only close notifications that are currently displayed (have an active widget)
        if (!notification.closed_at.has_value() && hasActiveWidget && hasActiveWidget(id)) {
            CloseNotification(id);
        }
    }
}

void NotificationService::OpenActiveNotifications() {
    logger.info("OpenActiveNotifications called");

    // Exclude empty action keys when counting.
    // Behavior:
    // - one non-empty action => invoke it
    // - multiple non-empty actions => error
    std::vector<std::pair<uint, QString>> ids_to_invoke;

    for (auto& [id, notification] : notifications) {
        if (!notification.closed_at.has_value()) {
            std::vector<QString> non_empty_action_keys;
            for (const auto& [key, value] : notification.actions) {
                Q_UNUSED(value);
                if (key.isEmpty()) {
                    continue;
                }
                non_empty_action_keys.push_back(key);
            }

            if (non_empty_action_keys.empty()) {
                // No actions, skip
                continue;
            }

            QString action_key;
            if (non_empty_action_keys.size() == 1) {
                action_key = non_empty_action_keys[0];
            }

            if (!action_key.isEmpty()) {
                ids_to_invoke.push_back({id, action_key});
            } else {
                if (calledFromDBus()) {
                    sendErrorReply(QDBusError::Failed,
                                   QString("Notification id %1 has more than one action").arg(id));
                }
                logger.error(
                    QString("OpenActiveNotifications: Notification %1 has more than one action")
                        .arg(id));
                return;
            }
        }
    }

    // Invoke all collected actions
    for (const auto& [id, action_key] : ids_to_invoke) {
        emit ActionInvoked(id, action_key);
    }
}

QStringList NotificationService::GetCapabilities() {
    return {"action-icons", "actions",     "body",        "body-hyperlinks",
            "body-images",  "body-markup", "persistence", "sound"};
}

void NotificationService::GetServerInformation(QString& name, QString& vendor, QString& version,
                                               QString& spec_version) {
    name = "notification-tray";
    vendor = "github.com";
    version = "0.0.1";
    spec_version = "1.3";
}
